vpt/sinks/sqlite_store/__init__.py

```diff
@@ -0,0 +1,91 @@
+import sqlite3
+import time
+from typing import Union
+
+import keyboard
+import mouse
+
+from vpt.sinks.base import SinkBase
+from vpt.sources.base import SourceBase
+
+
+class SQLiteStore(SinkBase):
+    '''Persistently store the engagement level and keyboard/mouse events
+       in an SQLite database.'''
+    connection: sqlite3.Connection
+
+    def __init__(self, db_path: str, mouse_source: SourceBase,
+                 keyboard_source: SourceBase, engagement_source: SourceBase):
+        '''Create a database or open an existing one.'''
+        self.connection = sqlite3.connect(db_path)
+        c = self.connection.cursor()
+        c.execute('''
+            CREATE TABLE IF NOT EXISTS engagement (
+                code integer NOT NULL,
+                timestamp integer NOT NULL
+            )
+        ''')
+        c.execute('''
+            CREATE TABLE IF NOT EXISTS keystrokes (
+                keycode integer NOT NULL,
+                modifiers text NOT NULL,
+                timestamp integer NOT NULL
+            )
+        ''')
+        c.execute('''
+            CREATE TABLE IF NOT EXISTS mouse_events (
+                type text NOT NULL,
+                x real,
+                y real,
+                delta integer,
+                button text,
+                timestamp integer NOT NULL
+            )
+        ''')
+        self.connection.commit()
+        c.close()
+
+        mouse_source.get_data_stream().subscribe(self.store_mouse_event)
+        keyboard_source.get_data_stream().subscribe(self.store_key_event)
+        engagement_source.get_data_stream().subscribe(self.store_engagement)
+
+    def __del__(self):
+        '''Clean up resources.'''
+        self.connection.commit()
+        self.connection.close()
+
+    def store_engagement(self, code: int):
+        '''Store an instance of engagement.'''
+        c = self.connection.cursor()
+        c.execute('''
+            INSERT INTO engagement VALUES (?, ?)
+        ''', (code, int(time.time())))
+        self.connection.commit()
+        c.close()
+
+    def store_key_event(self, event: keyboard.KeyboardEvent):
+        '''Store a keypress with all of its modifiers.'''
+        c = self.connection.cursor()
+        c.execute('''
+            INSERT INTO keystrokes VALUES (?, ?, ?)
+        ''', (event.scan_code, ','.join(modifiers), int(event.time)))
```

Comment:
Here, so the field in the table is called `keycode` but we feeding it the scancode. Consider changing the naming. There's a difference between keycodes and the scancodes. The first one are determined by the OS and can change based on the different keyboard layouts, but scancodes refer to the physical keys on the keyboard.